### PR TITLE
[Feature] Generate Wallet from secret numbers

### PR DESCRIPTION
**Pull request by nixer89**
## Generate Wallet from Secret Numbers

- added "fromSecretNumbers" method to Wallet instance
- added tests for new "fromSercretNumbers method

### Context of Change

This PR introduces a new feature. Many people are using XUMM. XUMM offers the recovery key for the user as [Secret Numbers](https://github.com/XRPLF/XRPL-Standards/issues/15)
To make it easy for developers to use their XUMM managed account for their development work, a new Method, "Wallet.fromSercretNumbers" is introduces. This method generates a new Wallet instance by the given Secret Numbers.

Alternatively, the developer had to first derive the family seed / secret key from the Secret Numbers and use the seed to initiate the Wallet instance.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Test Plan

Tests can be found in the index.ts of the /test/Wallet directory

## Future Tasks

The documentation has to be updated / generated. I did not run the "npm run docgen" target because it did not just added my method but changed lots of other things in the docs as well. So maybe someone else can run it.

### CODE REVIEW

File: packages/xrpl/src/Wallet/index.ts
Comment (JST5000) [inline packages/xrpl/src/Wallet/index.ts:120]: Is secretNumbers needed here? (I don't see it used anywhere in the constructor)

* If it is needed, could you add `@param opts.secretNumbers` + a description to the documentation comment?

File: packages/xrpl/src/Wallet/index.ts
Comment (JST5000) [inline packages/xrpl/src/Wallet/index.ts:210]: Please add `@param opts.algorithm` + description (Should be a description in the other methods with it)

File: packages/xrpl/src/Wallet/index.ts
Comment (JST5000) [inline packages/xrpl/src/Wallet/index.ts:224]: Since some users might be using javascript and pass in something unexpected, it's probably worth adding an exception if the type doesn't match. 
```suggestion
    } else {
         throw new ValidationError("Unable to parse the given secret numbers.")
    }
```

File: packages/ripple-keypairs/src/index.ts
Comment (ledhed2222) [inline packages/ripple-keypairs/src/index.ts:28]: @nixer89 this is what fixed the build issue

File: packages/ripple-address-codec/src/xrp-codec.ts
Comment (ledhed2222) [inline packages/ripple-address-codec/src/xrp-codec.ts:209]: FYI added this to suppress a linter warning.

File: packages/ripple-address-codec/src/xrp-codec.ts
Comment (ledhed2222) [inline packages/ripple-address-codec/src/xrp-codec.ts:220]: FYI added this to suppress a linter warning.

File: packages/xrpl/package.json
Comment (ledhed2222) [inline packages/xrpl/package.json:31]: should this be included in the monorepo?

Comment (JST5000) [reply packages/xrpl/package.json:31]: Maybe wait to bring it up to the top level until another package needs it?

Comment (ledhed2222) [reply packages/xrpl/package.json:31]: no i mean, should `xrpl-secret-numbers`'s source be in this monorepo. not sure where else it's used or how big it is, but that might make sense

File: packages/xrpl/src/Wallet/index.ts
Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts:117]: indeed not needed. removed

File: packages/xrpl/src/Wallet/index.ts
Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts:207]: added param + description

File: packages/xrpl/package.json
Comment (nixer89) [reply packages/xrpl/package.json:31]: I am sure the XRPL-Labs team uses this lib in their XUMM app. I think @WietseWind is the maintainer?
For now I would leave it as "external" import and would not add it to the monorepo. but maybe Wietse thinks this would be a good idea :)

Comment (WietseWind) [reply packages/xrpl/package.json:31]: @ledhed2222 It's a fairly small package, and the (non dev) dependencies overlap with `ripple-keypairs` (brorand, ripple-keypairs)

File: package-lock.json
Comment (JST5000) [inline package-lock.json:3]: Was it intentional to delete 18,000 lines from the package-lock? (maybe a rebase is required?)

Comment (nixer89) [reply package-lock.json:3]: before making my latest changes I have merged the develop branch into my "secret numbers" branch. When I check the history of the package-lock.json file of the develop branch, there are also some big changes which might have caused this:

https://github.com/XRPLF/xrpl.js/commit/9f19d771d24fa186ae23d586f380725f53a59059#diff-053150b640a7ce75eff69d1a22cae7f0f94ad64ce9a855db544dda0929316519

https://github.com/XRPLF/xrpl.js/commit/2b424276344b2aa8b8b76d621500f4d9e1436663#diff-053150b640a7ce75eff69d1a22cae7f0f94ad64ce9a855db544dda0929316519

Comment (mvadari) [reply package-lock.json:3]: I believe this is happening because you're using a different version of npm. I'm on v7.20.3.

Comment (nixer89) [reply package-lock.json:3]: I just updated to npm 7.24.x and the package-lock.json was not changed after "npm ci"
I still  believe it is because I have merged the develop branch into my branch

File: packages/xrpl/package.json
Comment (ledhed2222) [reply packages/xrpl/package.json:31]: that would be a good reason to maintain it within the monorepo :)

IMO, the criteria for whether it should be part of the monorepo:

1) is the purpose/scope of the package overlapping with the monorepo's? in this case i think it might be
2) does it depend on other packages w/in this monorepo, potentially requiring work on it and `ripple-keypairs` simultaneously.

File: package-lock.json
Comment (ledhed2222) [reply package-lock.json]: `npm ci` doesn't change the package-lock, instead it reads from the package-lock for a faster install with deterministic package resolution. if the package-lock needs to be written, as in this case, you need `npm i`

Comment (ledhed2222) [reply package-lock.json]: check out my documentation in `CONTRIBUTING.md` for how to add/remove packages in this repo

File: packages/xrpl/src/Wallet/index.ts
Comment (ledhed2222) [inline packages/xrpl/src/Wallet/index.ts:222]: unnecessary use of `let` instead of `const` is a pet peeve of mine lol. you can rewrite this like the following to not require a let:


```javascript
const numbersArray: string[] = []

if (typeof secretNumbers === 'string') {
  numberArray.concat(Utils.parseSecretString(secretNumbers))
} else if (Array.isArray(secretNumbers)) {
  numbersArray.concat(secretNumbers)
} ....
```

**Review by JST5000 — DISMISSED** [review]
LGTM

File: packages/xrpl/src/Wallet/index.ts
Comment (ledhed2222) [inline packages/xrpl/src/Wallet/index.ts:228]: should use `DEFAULT_ALGORITHM` as a fallback here, not `secp256k1`

Comment (ledhed2222) [reply packages/xrpl/src/Wallet/index.ts:228]: you can also just forward the `opts` dict entirely:

```javascript
return Wallet.fromEntropy(entropy, opts)
```

File: packages/xrpl/src/Wallet/index.ts
Comment (ledhed2222) [reply packages/xrpl/src/Wallet/index.ts:222]: alternatively, there's a typescript syntax that lets you use `let` but also tell typescript that you expect the variable to never _actually_ be nullish. IMO this is strange to look at, so i prefer the above with `const` but another option:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

```javascript
let numbersArray!: string []

if (typeof secretNumbers === 'string') {
  numbersArray = Utils.parseSecretString(secretNumbers)
} ....
```

File: packages/xrpl/src/Wallet/index.ts
Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts:228]: I decided to fallback to `secp256k1` because, the only wallet using "Secret Numbers" so far is XUMM. And they use `secp256k1` and not `ed25519`.

So if someone wants to create a xrpl.js "Wallet.fromSecretNumbers" he always have to define the algorithm `secp256k1` or else he ends up with a different address.

File: packages/xrpl/src/Wallet/index.ts
Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts:222]: used your first solution! :)

File: packages/xrpl/src/Wallet/index.ts
Comment (JST5000) [reply packages/xrpl/src/Wallet/index.ts]: That makes sense, could you add a comment with that context?

Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts]: Yes, will do tomorrow 🙂

Comment (ledhed2222) [reply packages/xrpl/src/Wallet/index.ts]: Sorry I don't agree. I think that XUMM in this case should override the default. We want this lib to be useful to anyone and we want people to default to ed25519

Comment (nixer89) [reply packages/xrpl/src/Wallet/index.ts]: I meant it the other way around: Users create a wallet in XUMM. This process gives them the "Secret Numbers" via `secp256k1`.

Now this XUMM user is also a developer. So he wants to build on XRPL, using xrpl.js and his Secret Numbers he got from XUMM. If `ed25519` is the default, the user always has to specify the algorithm when using the ".fromSecretNumbers" method.

What is the reason to always default to ed25519?

Comment (JST5000) [reply packages/xrpl/src/Wallet/index.ts]: The reason to default to ed25519 is that it's apparently got better performance. (See comment on the two algorithms here: https://xrpl.org/cryptographic-keys.html#signing-algorithms)

File: packages/xrpl/src/Wallet/index.ts
Comment (Realetive) [reply packages/xrpl/src/Wallet/index.ts:222]: @nixer89 , first solution will not change original `numbersArray` — `.concat()` does not change the existing arrays, but instead returns a new array. Example with `let` is right.

**Comment by mvadari** [comment]
Should this PR be updated now?

**Comment by intelliot** [comment]
@nixer89 will you be working on this?


(Alternatively - please let us know if this is something that you would like to ask RippleX to pick up instead)

**Comment by mvadari** [comment]
> @nixer89 will you be working on this?
> 
> (Alternatively - please let us know if this is something that you would like to ask RippleX to pick up instead)

Isn't this already implemented?

**Comment by ckniffen** [comment]
> > @nixer89 will you be working on this?
> > (Alternatively - please let us know if this is something that you would like to ask RippleX to pick up instead)
> 
> Isn't this already implemented?

I know it was added to `xrpl-p@ 2.0`.

**Comment by mvadari** [comment]
> I know it was added to `xrpl-py@2.0`.

Whoops, I'm confusing Python and JS

File: packages/xrpl/test/wallet/index.test.ts
Comment (JST5000) [inline packages/xrpl/test/wallet/index.test.ts:357]: This test fails because of this issue: https://github.com/XRPLF/xrpl.js/pull/2376

Commented with a TODO to re-add the test once that PR is resolved. (Open to other ideas about what's the best way to handle this :) )

**Comment by ckniffen** [comment]
Ideally I would like to see this as a function separate from `Wallet` class.  Putting so much on the `Wallet` makes it impossible to tree shake.

**Comment by JST5000** [comment]
> Ideally I would like to see this as a function separate from `Wallet` class. Putting so much on the `Wallet` makes it impossible to tree shake.

@ckniffen I think changing this to not be part of `Wallet` should be part of a bigger set of breaking changes to refactor how we handle Wallet generation as a whole. With our current setup, I think it makes sense to add this to the `Wallet` class because that's where all generators are currently. 

Do you think it's worth having this be off the `Wallet` class before a more sweeping change?

File: packages/xrpl/.eslintrc.js
Comment (xrpl-grants) [inline packages/xrpl/.eslintrc.js:37]: What made this start becoming a problem?  This should be moved to a new PR.

**Review by ckniffen — APPROVED** [review]
As long as the other PR gets merged first and the tests are corrected I am good

**Review by jonathanlei — APPROVED** [review]

**Pull request merged**
